Override `to_json` behavior introduced by plugin gems (ASC-36)

ssciolla opened the pull request:
This PR aims to resolve [ASC-36](https://mlit.atlassian.net/jira/software/projects/ASC/boards/88?selectedIssue=ASC-36).

- [x] Add patch
- [x] Add `rspec` setup using Docker, GitHub Actions

Resources
- https://stackoverflow.com/questions/683989/how-do-you-deal-with-the-conflict-between-activesupportjson-and-the-json-gem
- https://github.com/rails/rails/blob/v5.2.8/activesupport/lib/active_support/core_ext/object/json.rb
- https://bundler.io/guides/groups.html#optional-groups-and-bundlewith

ssciolla committed:
Add patch from Roger to backend init

ssciolla committed:
Add more explanation in comments

ssciolla committed:
Add one more sentence

ssciolla committed:
Add rspec with basic unit test?

ssciolla committed:
Tweak libraries to make testing correct

ssciolla committed:
Add some local Docker testing artifacts

ssciolla committed:
Tidy up Docker artifacts

ssciolla committed:
Switch for now to CRuby

ssciolla committed:
Add a comment to Gemfile; tweak spec strings

ssciolla committed:
Add newline at EOF in .dockerignore

ssciolla committed:
Add a basic README file

ssciolla commented on Dockerfile:
```diff
@@ -0,0 +1,28 @@
+
+COPY --chown=${UID}:${GID} . /app
+
+RUN bundle config set --local with test
+RUN bundle install
```
I opted to just include the gems in the Docker image, since there seemed to be some kind of issue preventing me from rebuilding with `docker compose run test bundle install` anyway.

niquerio replied:
That's fine.

ssciolla commented on Dockerfile:
```diff
@@ -0,0 +1,28 @@
+FROM ruby:2.5.9
```
I am not currently using JRuby, since building with the equivalent version (`jruby:9.2.20.1-jre11`) seems to take a really long time and does not always finish. I believe I got it to work once, and will keep trying, but I feel like this should be sufficient for now, since I reproduced the particular issue with CRuby as well.

niquerio replied:
I agree! The fact that it's doing the same thing in CRuby is good enough and so much faster.

niquerio commented on backend/plugin_init.rb:
```diff
@@ -0,0 +1,15 @@
+# Patch ActiveSupport to_json implementation so it uses the json gem version of the method
+# that is expected by the Sinatra backend. This patch counteracts the effect of ActiveSupport
+# which was introduced by multiple gems needed by this plugin.
+
```
This is all great.

You might consider adding some context about where it breaks in ArchivesSpace, so someone in the future could more easily check if this is still needed. Especially since the test is a couple steps removed from the ASpace enviornment. (i.e. unit test in c ruby)

niquerio replied:
So, basically something says "making a report will demonstrate the bug".

ssciolla replied:
Sure, done, see https://github.com/mlibrary/aspace-oauth/pull/5/commits/ecb544b7ef4fe9150212322f7a2013608e6e711d

niquerio commented on spec/backend_plugin_init_spec.rb:
```diff
@@ -0,0 +1,19 @@
+require "date"
+require "logger"
+
+require "json"
```
Just adding that I ran this locally, and the test fails and succeeds in the same way with or without this line. 

So I'd be inclined to remove this from here and the Gemfile, since maybe using the `json` gem isn't the issue. But it's up to you.

ssciolla replied:
As discussed, I'm keeping this cause if you don't `require` JSON or `active_support` here, there is no implementation of `to_json`. I added some comments to the imports, see https://github.com/mlibrary/aspace-oauth/pull/5/commits/ecb544b7ef4fe9150212322f7a2013608e6e711d.

niquerio reviewed (commented):
This all looks really good. You're well set up to add more tests in the future if you need to. 

The only other thing I might add is a github action to run tests. This is an example: https://github.com/mlibrary/account/blob/main/.github/workflows/tests.yml

You can get rid of everything in the tests workflow about environment variables, and you'll have to change the ruby version.

You also might want to add standard and then that part of the tests workflow will work.

I don't remember if you've used github actions much, but in case not, that workflow won't run until it exists in the default branch.

ssciolla committed:
Add some more explanatory comments

ssciolla committed:
Add GitHub workflow for tests

ssciolla committed:
Modify versions; run bundle install manually

ssciolla commented on .github/workflows/tests.yaml:
```diff
@@ -0,0 +1,20 @@
+      - uses: actions/checkout@v3
+      - name: Set up Ruby 2
+        uses: ruby/setup-ruby@v1
+        with:
+          ruby-version: '2.5.9'
+          rubygems: '3.2.3'
+          bundler: '2.3.26'
+      - name: Install gems
+        run: bundle config set --local with test && bundle install
```
Had to modify all this to do the conditional install of `rspec`/`json`. This is basically all documented in `Dockerfile`; do you ever use a Docker build action for this instead?

ssciolla replied:
It's working now: https://github.com/ssciolla/aspace-oauth/actions/runs/6356634768/job/17266580842

niquerio replied:
I don't usually do the conditional install for tests. Just install everything with the `bundler-cache: true` option.

for "using a Docker build action", are you thinking use `docker compose` to run tests? If so, then yes I have done that but I find it too slow.

niquerio replied:
It looks like you could set some environment variables to get the `with test` to work: https://github.com/ruby/setup-ruby#bundle-config

I think this should work:
```
jobs:
  # Run tests
  test:
    runs-on: ubuntu-latest
    env:
      - BUNDLE_WITH: test
    steps:
      - uses: actions/checkout@v3
      - name: Set up Ruby 2
        uses: ruby/setup-ruby@v1
        with:
          ruby-version: '2.5.9'
          rubygems: '3.2.3'
          bundler: '2.3.26'
          bundler-cache: true
      - name: Run tests
        run: bundle exec rspec
 ```

ssciolla replied:
Oh, nice, I can do that instead.

Do you not think it's necessary to exclude `rspec` and `json`? I guess this is part of my tendency to keep production images lean, which we've discussed before.

ssciolla replied:
Okay, looks like I got that changed.

ssciolla replied:
With Docker build, I was thinking instead of a `ruby/setup-ruby` step, there'd be something that builds a Docker image that we could then run in a subsequent step (maybe enabled by docker compose, maybe not). Just a thought, since that's how we're testing locally. You are correct that this is probably faster!

ssciolla commented:
> The only other thing I might add is a github action to run tests.

Done, see https://github.com/mlibrary/aspace-oauth/pull/5/commits/b113f95ea439aa63922b75173a1e660db6932871

niquerio reviewed (commented):
There are some thoughts on the tests.yaml

ssciolla committed:
Use BUNDLE_WITH and bundler-cache

ssciolla committed:
Tweak YAML syntax

ssciolla commented:
Hi @niquerio, usually I'd wait for the PR to be approved, but I think I've addressed your comments, and that you're OK with the fix and other additions. I'll follow up with you to see if there's anything else you'd recommend changing later on.
